### ISSUE TITLE
Generate .proto file from desc.FileDescriptor

### ISSUE THREAD

**Issue by jkinkead** [jhump/protoreflect#66]
Feature request. This may exist in a place I haven't seen yet.

It would be cool to be able to dynamically create or modify a FileDescriptor, then write out the new .proto definition.

This would be even cooler if protoparse-generated descriptors included source_code_info so that round-tripping modifications were possible without losing comment metadata.

**Comment by jhump** [comment]
I have thought about writing out proto files from a descriptor. I agree it would be a neat feature.

Making it lossless, for cases where `source_code_info` *is* present, is actually very complicated because of the way the `source_code_info` models language elements and file locations. Also, the way extra comments are represented, like inline comments between elements, makes it really hard to preserve *all* comments in the output. For example:

```proto
message Foo {
  /* comment 1 */ optional /* comment 2 */ /* comment 3 */ string
    // yet another comment
    fieldNameBar = /* more comments */ 3 [ /* some option comments! */ default = 
    // another line comment, inside options
     "baz" ]/* another one! */ ; // trailer
}
```

I believe there is enough info in `source_code_info` to be aware of all of the comments, but not enough to actually preserve formatting and placement.

----

I have seriously looked at preserving comments in the `protoparse` package. It is a hefty change that would involve quite a lot of boiler-plate -- wrapping every value in the parse tree with a new struct that includes position information and comments. There would also be some non-trivial work to make sure comments are attributed to elements correctly (to match protoc behavior).

I'll add this as a separate issue.

**Comment by jhump** [comment]
Feature implemented in #84
